File: _Graph/graph.py
from edge import *

class Graph:
    def __init__(self) -> None:
        self.__adjacency_list: dict[Vertex, set[Edge]] = {}
        
    def get_adjacency_list(self) -> dict[Vertex, set[Edge]]:
        return self.__adjacency_list
    
    def get_vertices(self) -> set[Vertex]:
        return set(self.__adjacency_list.keys())
    
    def add_vertex(self, label: str, heuristic: int) -> Vertex:
        vertex: Vertex = Vertex(label, heuristic)
        self.__adjacency_list[vertex] = set()
        return vertex
    
    def get_vertex(self, label: str) -> Vertex | None:
        for vertex in self.get_vertices:
            if vertex.label == label: return vertex
        return None
    
    def add_edge(self, u: Vertex, v: Vertex, w: int) -> Edge:
        edge: Edge = Edge(u, v, w)
        self.__adjacency_list[u].add(edge)
        return edge
    
    def __str__(self) -> str:
        s: str = ''
        for vertex in self.get_vertices():
            s += str(vertex) + ' : ' + str(self.__adjacency_list[vertex]) + '\n'
        return s

File: _Graph/graph.cpp
#include "edge.cpp"
#include <bits/stdc++.h>

using namespace std;

class Graph {
    map<Vertex*, set<Edge*>> adjacencyList;
    set<Vertex*> vertices;

    public:
    inline map<Vertex*, set<Edge*>> getAdjacencyList() const;
    inline set<Vertex*> getVertices() const;
    Vertex* addVertex(string label, int heuristic);
    optional<Vertex*> getVertex(string label) const;
    Edge* addEdge(Vertex* u, Vertex* v, int w);
};

inline map<Vertex*, set<Edge*>> Graph::getAdjacencyList() const { return adjacencyList; }

inline set<Vertex*> Graph::getVertices() const { return vertices; }

Vertex* Graph::addVertex(string label, int heuristic) {
    Vertex* vertex = new Vertex(label, heuristic);
    vertices.insert(vertex);
    adjacencyList[vertex] = set<Edge*>();
    return vertex;
}

optional<Vertex*> Graph::getVertex(string label) const {
    for (Vertex* vertex : getVertices())
        if (vertex->label == label) return vertex;
    return nullopt;
}

Edge* Graph::addEdge(Vertex* u, Vertex* v, int w) {
    Edge* edge = new Edge(u, v, w);
    adjacencyList[u].insert(edge);
    return edge;
}
